feat: Add email registration page

rayrunzewang opened the pull request:
feat: add user registration backend

Added user registration backend:
- Implemented sign-up functionality with validation of incoming values.
- Added unit tests for the registration module and related functions.

Sign-up validation:
- Password: - Must contain 8-32 characters. - Must contain at least one letter and one number. - Must contain at least one special character.
- Display name: - Must contain 4-15 characters. - Only letters, numbers, dashes, and underscores are allowed. If it is empty, a default one will be provided.
- Reference name: - Currently, a pre-filled Reference Code is defined in environment variables in both the backend and frontend.

Resolved CP-5
![2024-04-20 10_09_40-Git History (platform_api) - platform_api - Visual Studio Code](https://github.com/BeeQuantAI/platform_api/assets/80236130/87d6145c-8a70-4ff2-bc73-f0f642a4fe92)

Chris-Wang commented on README.md:
```diff
@@ -25,6 +25,8 @@ DB_PORT=5432
 DB_PASSWORD=CORE_ADMIN
 JWT_SECRET=hello_beeQuant
+DEFAULT_REFERENCE=COREINTERNAL
+DEFAULT_DISPLAY_NAME=Default_Display_Name
```
better to have a more read friendly name, for display purpose

Chris-Wang commented on src/modules/auth/user.dto.ts:
```diff
@@ -0,0 +1,22 @@
+  password: Joi.string()
+    .pattern(
+      new RegExp(
+        '^(?=.*[A-Za-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*_+=\\\\~<>:;”’(),./[\\]_`|{}-])[A-Za-z0-9#?!@$%^&*_+=\\\\~<>:;”’(),./[\\]_`|{}-]{8,32}$',
```
consider extract regex to a helper file and import here

Chris-Wang commented on src/modules/user/user.resolver.ts:
```diff
@@ -7,7 +7,7 @@ import { SkipAuth } from '../auth/skip-auth.decorator';
 @Resolver()
 export class UserResolver {
-  constructor(private readonly userService: UserService) {}
+  constructor(private readonly userService: UserService) { }
```
?

Chris-Wang commented on .eslintrc.js:
```diff
@@ -21,5 +21,11 @@ module.exports = {
+    'prettier/prettier': [
+      'error',  
+      {
+        endOfLine: 'auto',   
```
I don't think this is necessary needed, if every time PC user has CRLF as endOfLine, and Mac user has LF as endOfLine, our repo will keep too format...

Chris-Wang replied:
You can think another solution, hint: vscode setting

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -1,83 +1,29 @@
-import * as bcrypt from 'bcryptjs';
-import { JwtService } from '@nestjs/jwt';
+// import { JwtService } from '@nestjs/jwt';
```
again, don't have commented code in a pr unless there is a good reason

Chris-Wang commented on src/modules/auth/auth.service.ts:
```diff
@@ -1,6 +1,85 @@
+        process.env.DEFAULT_DISPLAY_NAME || 'Default_Display_Name';
+    }
+
+    if (!input.email || !input.password || !input.ref || !input.displayName) {
```
since `ValidationPipe` is introduced in resolver, the logic of `register` service need no change?

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -1,83 +1,29 @@
+  constructor(private authService: AuthService) {}
 
-  @Mutation(() => Result, { description: 'User login' })
+  // @Mutation(() => Result, { description: 'User login' })
```
don't have commented code in a pr unless there is a good reason

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -1,83 +1,26 @@
-  ) {}
+  constructor(private authService: AuthService) {}
 
-  @Mutation(() => Result, { description: 'User login' })
```
why remove the description?

Chris-Wang commented on src/modules/auth/auth.resolver.ts:
```diff
@@ -1,83 +1,26 @@
   }
 
-  @Mutation(() => Result, { description: 'User register' })
+  @Mutation(() => Result)
```
lack description

Chris-Wang commented on src/modules/auth/user.dto.ts:
need to rename and put under src/validation/schemas/auth/user.request.ts

Chris-Wang reviewed (changes requested):
Resolve comments pls

rayrunzewang committed:
feat: add user registration backend

Added user registration backend:

Implemented sign-up functionality with validation of incoming values.
Added unit tests for the registration module and related functions.
Sign-up validation:

Password: - Must contain 8-32 characters. - Must contain at least one letter and one number. - Must contain at least one special character.
Display name: - Must contain 4-15 characters. - Only letters, numbers, dashes, and underscores are allowed. If it is empty, a default one will be provided.
Reference name: - Currently, a pre-filled Reference Code is defined in environment variables in both the backend and frontend.
Resolved CP-5

Chris-Wang reviewed (approved):
LGTM
